### ISSUE TITLE
Mechanism to specify Krustlet's Pod CIDR?

### ISSUE THREAD

**Issue by DazWilkin** [deislabs/krustlet#252]
> TL;DR Should the Krustlet include a flag to specify its Pod CIDR?

I'm on thin ice here, so feel free to close if this is a non-sensible request.

The Krustlet appears to always grab `10.244.0.0/24` as the CIDR for its pods.

```
kubectl get node/krustlet \
--output=jsonpath='{.spec.podCIDR}'
10.244.0.0/24
```

Running on (a single node) MicroK8s cluster, both nodes use the same CIDR range:

```
kubectl get nodes \
--output=jsonpath='{range .items[*]}{.metadata.name}{"\t"}{.spec.podCIDR}{"\n"}{end}'
hades-canyon	10.244.0.0/24
krustlet	10.244.0.0/24
```
Running on a multi-node GKE cluster, GKE shards a CIDR range over its nodes:

```
kubectl get nodes \
--output=jsonpath='{range .items[*]}{.metadata.name}{"\t"}{.spec.podCIDR}{"\n"}{end}'
gke-clepsydra-default-pool-4c7699d9-dtq1	10.52.1.0/24
gke-clepsydra-default-pool-c8f03ba5-39v8	10.52.0.0/24
gke-clepsydra-default-pool-e515bbeb-v5hc	10.52.2.0/24
krustlet-01	10.244.0.0/24
krustlet-02	10.244.0.0/24
krustlet-03	10.244.0.0/24
```
> **NOTE** The GKE nodes are sharded across `10.52.0.0/16`
> **NOTE** The Krustlets all use `10.244.0.0/24`

IIUC this could result in:

+ Pods on different nodes colliding over the same PodIP?
+ Routing conflicts

IIUC on Google Cloud Platform, each node (Krustlet) joining the Kubernetes cluster, creates routes in the underlying network fabric to enable the platform to route traffic correctly. With one Krustlet, it has a unique CIDR (`10.244.0.0/24`) and so I think all's OK. But, with multiple Krustlets joining, I think there's an issue:

```
gcloud compute routes list \
--filter=description="k8s-node-route" \
--format="table(destRange,nextHopInstance.scope())" \
--project=${PROJECT}
```
yields:
```
DEST_RANGE     NEXT_HOP_INSTANCE
10.52.0.0/24   us-west1-a/instances/gke-clepsydra-default-pool-c8f03ba5-39v8
10.52.1.0/24   us-west1-c/instances/gke-clepsydra-default-pool-4c7699d9-dtq1
10.52.2.0/24   us-west1-b/instances/gke-clepsydra-default-pool-e515bbeb-v5hc
10.244.0.0/24  us-west1-a/instances/krustlet-01
```

The regular GKE nodes have unique Pod CIDRs and so they have unique routes. The Krustlets will attempt to share `10.244.0.0/24` and collide!?

I think it would be useful|necessary to include a flag to define the Pod CIDR for the node:

```
--pod-cidr="10.52.3.0/24"
```

**Comment by bacongobbler** [comment]
https://github.com/deislabs/krustlet/blob/6d2cbfe119cf893761a09c527189dfbec29f5177/crates/kubelet/src/node.rs#L597

**Comment by itowlson** [comment]
Is this something we want to do?  If so, I can take it.

**Comment by bacongobbler** [comment]
Yeah. As @DazWilkin pointed out, while this may not affect Krustlet itself, it will affect routing within certain cloud providers.

It looks now that one can set the pod CIDR by using [`set_pod_cidr`](https://github.com/deislabs/krustlet/blob/6d2cbfe119cf893761a09c527189dfbec29f5177/crates/kubelet/src/node.rs#L464). This can be done within a Provider by overriding [`node`](https://github.com/deislabs/krustlet/blob/9439b957c316945672b5c94b19add39d9773aa7d/crates/kubelet/src/provider/mod.rs#L98-L101). This has not been implemented by the [wasmtime](https://github.com/deislabs/krustlet/blob/9439b957c316945672b5c94b19add39d9773aa7d/crates/wasi-provider/src/lib.rs#L142-L147) or [wascc](https://github.com/deislabs/krustlet/blob/6d2cbfe119cf893761a09c527189dfbec29f5177/crates/wascc-provider/src/lib.rs#L195-L199) providers, however that can be achieved if we introduce a new config parameter to set the pod CIDR and read from that config value inside `.node`.

**Comment by lfrancke** [comment]
FYI for others stumbling across this:
`set_pod_cidr` doesn't take an Option but `podCIDR` is optional in Kubernetes.
By specifying an empty string one can still get it to work but that's not a very Rusty API I think.

We found this issue because this can bring down K8s Clusters (the control plane) down: https://github.com/kubernetes/kubernetes/issues/95948

I have looked at the virtual-kubelet and it doesn't seem to set `podCIDR` at all.
I'd say it makes sense to change the Krustlet default behavior as well.

**Comment by soenkeliebau** [comment]
Paging @pires just in case he has any comments on the virtual-kubelet behavior.

**Comment by pires** [comment]
[VK has no opinion when it comes to Pod CIDR](https://github.com/virtual-kubelet/virtual-kubelet/issues/164#issuecomment-388463236). A provider or the kube-controller-manager can specify the respective Node attribute.
